qa/L0_lifecycle/lifecycle_test.py

```diff
@@ -2625,6 +2626,112 @@ def test_load_gpu_limit(self):
+        # use a tighter bound of 15 seconds
+        self.assertLess(loading_time, 15.0,
+                        "Concurrent loading speedup not observed")
+        # Concurrent loading time cannot be < 10 seconds
+        self.assertGreaterEqual(loading_time, 10.0,
+                                "Invalid concurrent loading time")
+        # Make sure the models are loaded
+        self.assertTrue(triton_client.is_server_live())
+        self.assertTrue(triton_client.is_server_ready())
+        for model_name in model_names:
+            self.assertTrue(triton_client.is_model_ready(model_name))
+
+    def test_concurrent_load(self):
+        # Initialize client
+        try:
+            triton_client = grpcclient.InferenceServerClient("localhost:8001",
+                                                             verbose=True)
+        except Exception as ex:
+            self.assertTrue(False, "unexpected error {}".format(ex))
+        # Load same model concurrently
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            thread_1 = pool.submit(triton_client.load_model,
+                                   "identity_zero_1_int32")
+            time.sleep(2)  # wait between loads
+            thread_2 = pool.submit(triton_client.load_model,
+                                   "identity_zero_1_int32")
+            # Second load should wait until the first completes
+            # Both loads should succeed
+            thread_1.result()
+            thread_2.result()
+        self.assertTrue(triton_client.is_server_live())
+        self.assertTrue(triton_client.is_server_ready())
+        self.assertTrue(triton_client.is_model_ready("identity_zero_1_int32"))
+
+    def test_concurrent_load_unload(self):
+        # Initialize client
+        try:
+            triton_client = grpcclient.InferenceServerClient("localhost:8001",
+                                                             verbose=True)
+        except Exception as ex:
+            self.assertTrue(False, "unexpected error {}".format(ex))
+        # Load identity_zero_1_int32 and unload it while loading
+        # The unload operation should wait until the load is completed
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            load_thread = pool.submit(triton_client.load_model,
+                                      "identity_zero_1_int32")
+            time.sleep(2)  # wait between load and unload
+            unload_thread = pool.submit(triton_client.unload_model,
+                                        "identity_zero_1_int32")
+            load_thread.result()
+            unload_thread.result()
+        self.assertTrue(triton_client.is_server_live())
+        self.assertTrue(triton_client.is_server_ready())
+        self.assertFalse(triton_client.is_model_ready("identity_zero_1_int32"))
+        # Load ensemble_zero_1_float32 and unload its dependency while loading
+        # The unload operation should wait until the load is completed
```

Comment:
This is not checked in the test. And is this the requirement? i.e. without knowing implementation detail, should the test expect the unload to be blocked until the previous load return, or it just expect the model is marked unavailable after both operation completed (return order not enforced)

Reply:
> without knowing implementation detail, should the test expect the unload to be blocked until the previous load return

Yes. When load/unload requests are having conflict, any ongoing load/unload will block any incoming load/unload until the ongoing load/unload returns.

However, if there are >2 requests, i.e. the 1st is ongoing and the 2nd and 3rd are blocked, it is undefined whether the 2nd or 3rd request will be the next ongoing request after the 1st request returns. Thus, if the 1st is load, the 2nd is unload and 3rd is load. There are two possible execution orders: `load -> unload -> load` or `load -> load -> unload`, which the eventual state is unknown in advance. Therefore, if the test is going to cover >2 concurrent load/unload requests, the eventual model state is valid on both load or unload state.